src/io.c

```diff
@@ -72,76 +66,40 @@ uint32_t memory_read_str(memory_t *m, uint8_t *dst, uint32_t addr, uint32_t max)
-    assert(c);
-    return *(const uint32_t *) (c->data + addr_lo);
-}
-
-uint32_t memory_read_w(memory_t *m, uint32_t addr)
+uint32_t memory_ifetch(uint32_t addr)
 {
-    const uint32_t addr_lo = addr & mask_lo;
-    chunk_t *c = m->chunks[addr >> 16];
-    return *(const uint32_t *) (c->data + addr_lo);
+    return *(const uint32_t *) (data_memory_base + addr);
 }
 
-uint16_t memory_read_s(memory_t *m, uint32_t addr)
-{
-    const uint32_t addr_lo = addr & mask_lo;
-    chunk_t *c = m->chunks[addr >> 16];
-    return *(const uint16_t *) (c->data + addr_lo);
-}
+#define MEM_READ_IMPL(size, type)                   \
+    type memory_read_##size(uint32_t addr)          \
+    {                                               \
+        return *(type *) (data_memory_base + addr); \
+    }
 
-uint8_t memory_read_b(memory_t *m, uint32_t addr)
-{
-    chunk_t *c = m->chunks[addr >> 16];
-    return c->data[addr & mask_lo];
-}
+MEM_READ_IMPL(w, uint32_t);
+MEM_READ_IMPL(s, uint16_t);
+MEM_READ_IMPL(b, uint8_t);
 
-void memory_write(memory_t *m, uint32_t addr, const uint8_t *src, uint32_t size)
+void memory_write(memory_t *mem,
+                  uint32_t addr,
+                  const uint8_t *src,
+                  uint32_t size)
 {
-    for (uint32_t i = 0; i < size; ++i) {
-        const uint32_t addr_lo = (addr + i) & mask_lo,
-                       addr_hi = (addr + i) >> 16;
-        chunk_t *c = m->chunks[addr_hi];
-        if (!c) {
-            c = mpool_calloc(mp);
-            m->chunks[addr_hi] = c;
-        }
-        c->data[addr_lo] = src[i];
-    }
+    memcpy(mem->mem_base + addr, src, size);
 }
 
-#define MEM_WRITE_IMPL(size, type)                                           \
-    void memory_write_##size(memory_t *m, uint32_t addr, const uint8_t *src) \
-    {                                                                        \
-        const uint32_t addr_lo = addr & mask_lo, addr_hi = addr >> 16;       \
-        chunk_t *c = m->chunks[addr_hi];                                     \
-        if (unlikely(!c)) {                                                  \
-            c = mpool_calloc(mp);                                            \
-            m->chunks[addr_hi] = c;                                          \
-        }                                                                    \
-        *(type *) (c->data + addr_lo) = *(const type *) src;                 \
+#define MEM_WRITE_IMPL(size, type)                                 \
+    void memory_write_##size(uint32_t addr, const uint8_t *src)    \
+    {                                                              \
+        *(type *) (data_memory_base + addr) = *(const type *) src; \
     }
 
 MEM_WRITE_IMPL(w, uint32_t);
 MEM_WRITE_IMPL(s, uint16_t);
 MEM_WRITE_IMPL(b, uint8_t);
 
-void memory_fill(memory_t *m, uint32_t addr, uint32_t size, uint8_t val)
+void memory_fill(memory_t *mem, uint32_t addr, uint32_t size, uint8_t val)
```

Comment:
I am not sure if it still makes sense to keep `memory_fill` as a public function. Can you check?

Reply:
Alternatively, you may define it as a static inline function.

Reply:
memory_fill is a public function for `elf_load`.